umodbus/asynchronous/common.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2019, Pycom Limited.
+#
+# This software is licensed under the GNU GPL version 3 or any
+# later version, with permitted additional terms. For more information
+# see the Pycom Licence v1.0 document supplied with this file, or
+# available at https://www.pycom.io/opensource/licensing
+#
+
+# system packages
+from ..sys_imports import List, Optional, Tuple, Union
+
+# custom packages
+from .. import functions, const as Const
+from ..common import CommonModbusFunctions, Request
+
+
+class AsyncRequest(Request):
+    """Asynchronously deconstruct request data received via TCP or Serial"""
+
+    async def send_response(self,
+                            values: Optional[list] = None,
+                            signed: bool = True) -> None:
+        """
+        Send a response via the configured interface.
+
+        :param      values:  The values
+        :type       values:  Optional[list]
+        :param      signed:  Indicates if signed values are used
+        :type       signed:  bool
+        """
+        await self._itf.send_response(self,
+                                      self.unit_addr,
+                                      self.function,
+                                      self.register_addr,
+                                      self.quantity,
+                                      self.data,
+                                      values,
+                                      signed)
+
+    async def send_exception(self, exception_code: int) -> None:
+        """
+        Send an exception response.
+
+        :param      exception_code:  The exception code
+        :type       exception_code:  int
+        """
+        await self._itf.send_exception_response(self,
```

Comment:
not sure whether this really works and requires the `self` parameter. I also recommend to use named parameters

Reply:
It worked when I tried it with the async TCP implementation on Pycopy, which should be similar enough to Micropython at this stage that it would most likely work there too.

Basically, the interface of an `AsyncRequest` will always be the async implementation of the TCP/RTU class (e.g. `tcp.py@214:4`'s `AsyncTCP.send_response`), and those classes' `send_*response` functions take the `AsyncRequest` itself as a parameter (hence the `self` as the first argument). Since `AsyncRequest`s will never be used with synchronous classes, and since `Request`s will never be used (directly) with asynchronous classes, it shouldn't result in an error.

I'll change the function calls to use named parameters.

Reply:
On second thought, you're right - it won't work with the async Serial implementation because the `request` parameter is missing there. I've added it there for compatibility and marked it as `ignored` in the docstring.